pkg/services/provisioning/values/values_test.go

```diff
@@ -0,0 +1,169 @@
+package values
+
+import (
+	. "github.com/smartystreets/goconvey/convey"
+	"gopkg.in/yaml.v2"
+	"testing"
+)
+
+func TestValues(t *testing.T) {
+	Convey("Values", t, func() {
+		oldGetEnv := getEnv
+		getEnv = func(key string) string {
+			return map[string]string{
+				"INT":    "1",
+				"STRING": "test",
+				"BOOL":   "true",
+			}[key]
+		}
+
+		Convey("IntValue", func() {
+			type Data struct {
+				Val IntValue `yaml:"val"`
+			}
+			d := &Data{}
+
+			Convey("Should unmarshal simple number", func() {
+				unmarshalingTest(`val: 1`, d)
+				So(d.Val.Value(), ShouldEqual, 1)
+			})
+
+			Convey("Should unmarshal env var", func() {
+				unmarshalingTest(`val: $INT`, d)
+				So(d.Val.Value(), ShouldEqual, 1)
+			})
+
+			Convey("Should ignore empty value", func() {
+				unmarshalingTest(`val: `, d)
+				So(d.Val.Value(), ShouldEqual, 0)
+			})
+		})
+
+		Convey("StringValue", func() {
+			type Data struct {
+				Val StringValue `yaml:"val"`
+			}
+			d := &Data{}
+
+			Convey("Should unmarshal simple string", func() {
+				unmarshalingTest(`val: test`, d)
+				So(d.Val.Value(), ShouldEqual, "test")
+			})
+
+			Convey("Should unmarshal env var", func() {
+				unmarshalingTest(`val: $STRING`, d)
+				So(d.Val.Value(), ShouldEqual, "test")
+			})
+
+			Convey("Should ignore empty value", func() {
+				unmarshalingTest(`val: `, d)
+				So(d.Val.Value(), ShouldEqual, "")
+			})
+		})
+
+		Convey("BoolValue", func() {
+			type Data struct {
+				Val BoolValue `yaml:"val"`
+			}
+			d := &Data{}
+
+			Convey("Should unmarshal bool value", func() {
+				unmarshalingTest(`val: true`, d)
+				So(d.Val.Value(), ShouldBeTrue)
+			})
+
+			Convey("Should unmarshal explicit string", func() {
+				unmarshalingTest(`val: "true"`, d)
+				So(d.Val.Value(), ShouldBeTrue)
+			})
+
+			Convey("Should unmarshal env var", func() {
+				unmarshalingTest(`val: $BOOL`, d)
+				So(d.Val.Value(), ShouldBeTrue)
+			})
+
+			Convey("Should ignore empty value", func() {
+				unmarshalingTest(`val: `, d)
+				So(d.Val.Value(), ShouldBeFalse)
+			})
+		})
+
+		Convey("JSONValue", func() {
+
+			type Data struct {
+				Val JSONValue `yaml:"val"`
+			}
+			d := &Data{}
+
+			Convey("Should unmarshal variable nesting", func() {
+				doc := `
+                 val:
```

Comment:
Yaml does not seem to allow tabs and so these must be spaces which then creates odd indent.